Add WriteTo API on JsonDocument and JsonProperty

sasivishnu opened the pull request:
Fixes #39037 

Changes:

1. `WriteProperty` API in `JsonElement` is removed.
2. `WriteValue` API in `JsonElement` is renamed as `WriteTo`.
3. `WriteTo` API is added in `JsonDocument` and `JsonProperty`.

For unit tests, I referred previously available unit test methods for `JsonDocument` and `JsonProperty` new API `WriteTo`.

Thanks for all the help @bartonjs and @ahsonkhan . I have done as we discussed. Let me know if I missed anything.

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=39367) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=39367) before we can accept your contribution.<br/><br/>:x: sasivishnu [sign now](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=39367)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/corefx?pullRequest=39367) it.</sub>

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Document/JsonProperty.cs:
```diff
@@ -88,6 +88,25 @@ public bool NameEquals(ReadOnlySpan<char> text)
+        /// </exception>>
+        public void WriteTo(Utf8JsonWriter writer)
+        {
+            writer.WritePropertyName(Name);
```
I know this is an existing bug, but apparently we didn't add argument null validation to JsonElement's write methods when Utf8JsonWriter switched from a struct to a class.

Can you add the

```C#
if (writer == null)
    throw new ArgumentNullException(nameof(writer));
//blank line here that I can't express in this window
```
preamble to each of these WriteTo methods?

sasivishnu replied:
Yeah, we can do that. Need to update some test methods too since we are using `default` for UTF8JsonWriter. 

Since you mentioned it as bug, should I create separate PR to fix this for better tracking/information purpose?

bartonjs replied:
> should I create separate PR to fix this for better tracking/information purpose

Nah. The old WriteValue method had a bug, but you deleted it.  Let's make sure the new WriteTo method doesn't have the same bug :wink:.

sasivishnu replied:
Done. Added

bartonjs commented on src/System.Text.Json/tests/JsonDocumentTests.cs:
```diff
@@ -3740,6 +3671,451 @@ private static string GetCompactJson(TestCaseType testCaseType, string jsonStrin
+        [InlineData(true)]
+        public static void WriteNumber(bool indented)
+        {
+            WriteSimpleValue(indented, "42");
```
Rather than duplicate all of these tests it'd be better to make them reusable.  Given that we're on sort of a short timeline here I'd suggest just doing two tests for JsonDocument:

1) `AssertExtensions.Throws<ArgumentNullException>("writer", doc.WriteTo(null));`
2) One simple test that reads `"[ 1e400 ]"` and writes `"[1e400]"` using the default (non-indented) mode of the writer.


If you want to tackle reusability, that'd be
* Change `public static class JsonElementWriteTests` to `public abstract class JsonDomWriteTests`
* Add something like `public abstract void Write(JsonDocument document, Utf8JsonWriter writer);`
* Make WriteSimpleValue use the abstract method instead of `target.WriteValue(writer);`
* Maybe also add a `public abstract bool ShouldWrap { get; }`?
* Use ShouldWrap appropriately.
* Make two new types at the top of the file

```C#
public class JsonElementWriteTests : JsonDomWriteTests
{
    public override bool ShouldWrap => true;
    public override void Write(JsonDocument document, Utf8JsonWriter writer) => document.RootElement[0].WriteTo(writer);
}

public class JsonDocumentWriteTests : JsonDomWriteTests
{
    public override bool ShouldWrap => false;
    public override void Write(JsonDocument document, Utf8JsonWriter writer) => document.WriteTo(writer);
}
```

* Make sure the tests are actually still executing by making one fail (you might have to remove the word static from all of them once you're using inheritance... I forget)
* Fix the failure before pushing the new commits :)

bartonjs replied:
(and any tests which were specific to using WriteTo from `default(JsonElement)` would then get moved into JsonElementWriteTests)

sasivishnu replied:
I had other confusion too. Since `document.WriteTo` inturn calls `RootElement.WriteTo`, do we need to test all those scenario again? 

Going for reusability now might get complicated I think. I will work on this as separate PR if needed. For now, I will remove all those tests and add simple test as you mentioned. Is that fine?

bartonjs replied:
> For now, I will remove all those tests and add simple test as you mentioned. Is that fine?

Yep.

sasivishnu replied:
Done. Removed those tests and added one simple test for now.

bartonjs commented on src/System.Text.Json/tests/JsonPropertyTests.cs:
```diff
@@ -0,0 +1,289 @@
+        [Fact]
+        public static void CheckUseAfterDispose()
+        {
+            using (JsonDocument doc = JsonDocument.Parse("{\"First\":1}", default))
```
Again, really this should reuse the tests; but then provide whatever extra tests we need that would be unique to JsonProperty.

I'd rather see these tests be white-box and minimal for the 3.0 release than see all of this code/logic duplicated.

sasivishnu replied:
> 
> 
> Again, really this should reuse the tests; but then provide whatever extra tests we need that would be unique to JsonProperty.
> 

Same here, I will move assertion in existing test methods itself. Not sure of unique test cases for JsonProperty but I will think about it.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs:
```diff
@@ -75,6 +75,29 @@ public void Dispose()
+        /// </summary>
+        /// <param name="writer"></param>
+        /// <exception cref="ArgumentNullException">
+        ///   Parameter <paramref name="writer"/> of type <see cref="Utf8JsonWriter"/> should not be null.
```
The standard boilerplate for this is more like

```suggestion
        ///   The <paramref name="writer"/> parameter is <see langword="null"/>.
```

(It describes when the exception is thrown, vs the guidance to avoid it)

sasivishnu replied:
corrected it.

bartonjs commented on src/System.Text.Json/tests/JsonDocumentTests.cs:
```diff
@@ -1834,21 +1842,28 @@ public static void CheckUseDefault()
-                root.WriteProperty(ReadOnlySpan<char>.Empty, writer);
-            });
+                var expectedErrorMessage = @"Value cannot be null. (Parameter 'writer')";
+                AssertExtensions.Throws<ArgumentNullException>(() => doc.WriteTo(null), expectedErrorMessage);
```
```suggestion
                AssertExtensions.Throws<ArgumentNullException>("writer", () => doc.WriteTo(null));
```

bartonjs commented on src/System.Text.Json/tests/JsonDocumentTests.cs:
```diff
@@ -1834,21 +1842,28 @@ public static void CheckUseDefault()
-                root.WriteProperty(ReadOnlySpan<byte>.Empty, writer);
-            });
+                JsonElement root = doc.RootElement;
+                AssertExtensions.Throws<ArgumentNullException>(() => root.WriteTo(null), expectedErrorMessage);
```
The tests here on the element and properties belong in JsonElement and JsonProperty -specific tests, they don't need to be repeated here.

sasivishnu replied:
Done. Removed here and moved into their respective files.

bartonjs commented on src/System.Text.Json/tests/JsonDocumentTests.cs:
```diff
@@ -3740,6 +3698,47 @@ private static string GetCompactJson(TestCaseType testCaseType, string jsonStrin
+            {
+                var options = new JsonWriterOptions
+                {
+                    Indented = false,
```
Since this is just specifying the default options it can be entirely gotten rid of.

sasivishnu replied:
done

bartonjs commented on src/System.Text.Json/tests/JsonDocumentTests.cs:
```diff
@@ -28,6 +28,18 @@ public static class JsonDocumentTests
+            };
+
+        private static readonly bool s_replaceNewlines =
+            !StringComparer.Ordinal.Equals(CompiledNewline, Environment.NewLine);
```
None of `CompiledNewline`, `s_options`, or `s_replaceNewlines` is used any longer, they can be removed.

sasivishnu replied:
removed dead code.

bartonjs commented on src/System.Text.Json/tests/JsonPropertyTests.cs:
```diff
@@ -0,0 +1,149 @@
+{
+    public static class JsonPropertyTests
+    {
+
```
Should add the null-writer => ArgumentNullException test here.

sasivishnu replied:
Added new test here

bartonjs commented on src/System.Text.Json/tests/JsonElementWriteTests.cs:
```diff
@@ -901,19 +905,31 @@ public static void WritePropertyOutsideObject(bool skipValidation)
+                            (ref Utf8JsonWriter w) =>
+                            {
+                                w.WritePropertyName(CharLabel);
+                            });
```
Since these are being updated already they can be simplified (since the writer is no longer a struct) to just be

```C#
Assert.Throws<InvalidOperationException>(() => w.WritePropertyName(CharLabel));
```

(etc).

sasivishnu replied:
changed in all places at JsonElementWriteTests.cs

bartonjs commented on src/System.Text.Json/tests/JsonPropertyTests.cs:
```diff
@@ -0,0 +1,149 @@
+using System.Collections;
+using System.Globalization;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
```
I don't think this file needs the Newtonsoft.Json using-imports; and probably can get rid of a lot of these... can we pare it down to the ones that matter?

bartonjs commented on src/System.Text.Json/tests/JsonElementWriteTests.cs:
```diff
@@ -901,19 +905,31 @@ public static void WritePropertyOutsideObject(bool skipValidation)
+                            });
                     }
 
                     writer.Flush();
```
The test for sending a null writer to JsonElement.WriteTo should be in this file.

ericstj commented:
Let's fill out https://github.com/dotnet/docs/issues/new?template=doc-breaking-change.md to document the breaking change from Preview7.  @bartonjs or @sasivishnu can one of you do this?

sasivishnu commented:
> 
> 
> Let's fill out https://github.com/dotnet/docs/issues/new?template=doc-breaking-change.md to document the breaking change from Preview7. @bartonjs or @sasivishnu can one of you do this?

Yep, I can do it.

sasivishnu commented:
> 
> 
> Let's fill out https://github.com/dotnet/docs/issues/new?template=doc-breaking-change.md to document the breaking change from Preview7. @bartonjs or @sasivishnu can one of you do this?

@ericstj  I read the document about what all considered as breaking change. Just to confirm before filing, listing here the changes made and do they count as breaking changes.

1. All WriteProperty API removed from `JsonElement` - **Breaking change**
2. Renamed API WriteValue as WriteTo in `JsonElement` - **Breaking change**
3. [Bug Fix] WriteTo API in JsonElement now throws `ArgumentNullException` when method parameter of type `UTF8JsonWriter` is null - **Breaking change**
4. Added new API WriteTo in `JsonDocument` and `JsonProperty` - **No**

bartonjs commented:
@sasivishnu I'd just do the removed/renamed methods.  The NullReferenceException -> ArgumentNullException I wouldn't bother with.  (Though if @ericstj disagrees, listen to him :smile:).

ericstj commented:
No need to doc NRE -> ANE.

ericstj commented:
I couldn't find any upstack callers to the changed methods on github.  FYI: @nguerrera @pranavkm

nguerrera commented:
+ @wli3

sasivishnu commented:
@ericstj Done dotnet/docs#13374

Although I removed WriteProperty and renamed WriteValue methods, microsoft [docs ](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonelement.writeasproperty?view=netcore-3.0) says JsonElement has WriteAsProperty and WriteAsValue methods.

ericstj commented:
I believe these changes will be automatically addressed when the docs pick up the new API data.  /cc @carlossanlop

carlossanlop commented:
@ericstj The automatic porting tool won't port any comments that are already documented in MS Docs. In this case, the modification will have to be done manually by the area owners.
But if we are talking about a structural change of the APIs, that means all the comments may go away and I could port them from scratch. Is that the case?

ericstj commented:
> Is that the case?

Yes I believe so.

carlossanlop commented:
@sasivishnu @ericstj 

FYI and as an example, I have a [PR out ](https://github.com/dotnet/dotnet-api-docs/pull/2737/files)for System.Text.Json docs that I submitted 3 days ago.

Whatever comments you add in this change, I can port them to the Docs repo after we make sure the newest Json APIs are included in the Docs xml structure.
